src/session/content/message_row/document.rs

```diff
@@ -0,0 +1,280 @@
+            let imp = self.imp();
+
+            let message_text = parse_formatted_text(data.caption);
+            imp.content_label.set_visible(!message_text.is_empty());
+            imp.content_label.set_label(message_text);
+
+            imp.file_name.set_label(&data.document.file_name);
+
+            let session = message.chat().session();
+
+            self.update_status(data.document.document, session);
+        }
+    }
+
+    fn update_status(&self, document: File, session: Session) {
+        let local = &document.local;
+        let imp = self.imp();
+
+        if local.is_downloading_completed {
+            imp.file_status.set_icon_name("folder-documents-symbolic");
+            imp.file_size
+                .set_label(&format_size(local.downloaded_size as u64));
+
+            let path = local.path.clone();
+            let gio_file = gio::File::for_path(&path);
+
+            imp.file_status.connect_clicked(move |_| {
+                gio::AppInfo::launch_default_for_uri(
+                    &gio_file.uri(),
+                    Option::<&gio::AppLaunchContext>::None,
+                )
+                .ok();
+            });
+        } else if !local.is_downloading_active {
+            imp.file_status.set_icon_name("document-save-symbolic");
+
+            let file_size = if document.size != 0 {
+                document.size as u64
+            } else {
+                document.expected_size as u64
+            };
+
+            imp.file_size.set_label(&format_size(file_size));
+
+            let id = document.id;
+
+            imp.file_status.connect_clicked(
+                clone!(@weak self as obj, @strong session => move |file_status| {
+
+                    // I don't know how to cancel downloading
+                    file_status.set_icon_name("media-playback-stop-symbolic");
+                    file_status.connect_clicked(|_| {});
+
+                    let (sender, receiver) = glib::MainContext::sync_channel::<File>(Default::default(), 5);
+                    receiver.attach(
+                        None,
+                        clone!(@weak obj, @weak session => @default-return glib::Continue(false), move |file| {
+                            if file.local.is_downloading_completed {
+                                obj.update_status(file, session);
+                                glib::Continue(false)
+                            } else {
+                                let progress = file.local.downloaded_size as f64 / file.expected_size as f64;
+
+                                obj.imp().file_size.set_label(
+                                    format!(
+                                        "{:.0}% of {}",
+                                        progress * 100.0,
+                                        format_size(document.expected_size as u64)
+                                    )
+                                    .as_str(),
+                                );
```

Comment:
This sometimes causes the widget to grow for a split second when downloading. This is because it could happen that this text is larger than the one in the name label. Now, since I've planned to implement a radial progress bar inside the download widget to show progress later, I'd say it's better to drop this code for now so that we don't get that weird behavior.

Reply:
![image](https://user-images.githubusercontent.com/63008755/187511067-d3b0ae1b-7872-48d1-91f4-7af90c17512e.png)
looks like telegram reserves some place for that
we can add 10 spaces to regular status

Reply:
>  looks like telegram reserves some place for that

Makes sense, but to do that we need two things:

- The text should not be translatable, otherwise we're not sure of the max width of the label. To do that we can use the slash character instead of the word "of" in that string (e.g. "2 MB / 10 MB").
- We have to calculate the correct max width of the text. In this case using `glib::format_size` the longest producible text is probably something like "999.9 MB / 999.9 MB", which is 19 in total.

>  we can add 10 spaces to regular status

We can use `GtkLabel::width-chars` property to do that, which is cleaner. This property basically ensures that the label is always sized to fit the given number of character without resizing. I've tested to set this to 19, but I've noticed that it leaves a bit of extra space at the end (this is because some chars are thinner and I guess the width calculation evaluates the worst case scenario), so I think 18 is a more fine value in our case.

Reply:
I've updated message loading style like this
![Screenshot from 2022-09-03 19-19-47](https://user-images.githubusercontent.com/63008755/188279842-9e6f95b1-653f-4ad5-aeff-4acbfc4bc3c2.png)
![Screenshot from 2022-09-03 19-20-02](https://user-images.githubusercontent.com/63008755/188279839-d300e06a-317d-4501-a5e1-b4e3bb862c1c.png)
but i prefer style when loading hide size because it's require less space
![Screenshot from 2022-09-03 19-31-43](https://user-images.githubusercontent.com/63008755/188279889-03ecdd55-6be9-4e4a-bddf-ef8422b56d8c.png)